### PR TITLE
group constraints by arg

### PR DESCRIPTION
**Pull request by avikchaudhuri**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #101815

Before, we would emit a soup of specializations / constraints without any obvious order to guide readability.

With this diff, we group such results by arg, and add comments preceding each group. Empirically, the results read much better.

Differential Revision: [D45995199](https://our.internmc.facebook.com/intern/diff/D45995199/)

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by avikchaudhuri** [comment]
Failing tests would be fixed by https://github.com/pytorch/pytorch/pull/100739

File: test/test_dynamic_shapes.py [hunk 1]
```diff
@@ -1772,31 +1772,55 @@ def test_dim_constraints_solve_full(self):
...
+        def dummy_f(x0, x1, x2, x3, x4, x5, x6, x7, x8, x9, x10, x11, x12):
             pass
 
         action_code = dim_constraints.prettify_results(inspect.signature(dummy_f))
         static_code, dynamic_code = re.findall(r"```(.*?)```", action_code, re.DOTALL)
         print(static_code)
         expected_static = '''
-def specializations(x0, x1, x2, x3, x4, x5, x6, x7, x8, x9, x11, x12):
+def specializations(x0, x1, x2, x3, x4, x5, x6, x7, x8, x9, x10, x11, x12):
+    # x0:
```
Comment (angelayi) [inline test/test_dynamic_shapes.py:1783]: Could you add a test that would have multiple specializations on x0 to check the grouping? I would expect it to look something like:
```
# x0:
assert x0.size()[0] == <something>
assert x0.size()[1] == <something>
# x1:
assert x1.size()[0] == <something>
assert x1.size()[2] == <something>
```

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -1740,33 +1740,75 @@ def prettify_results(self, original_signature: inspect.Signature):
...
+            return arg, dc
+
+        def group(results, args_index):
+            groups = defaultdict(list)
+            for dc in results:
+                local = extract_and_rewrite_local(dc)
+                if local is None:
+                    # This can happen, e.g., with `assume_constant_result`.
+                    # In that case, we drop the constraint.
+                    # TODO(avik) Maybe we should generate an assertion here?
```
Comment (tugsbayasgalan) [inline torch/fx/experimental/symbolic_shapes.py:1758]: I feel like this is going to be hard as the value is completely baked.. One wild idea is we pass it as an extra input to the export call and assert it inside the graph?

**Review by tugsbayasgalan — APPROVED** [review]

File: test/test_dynamic_shapes.py (hunk 1 above)
Comment (avikchaudhuri) [reply test/test_dynamic_shapes.py]: Done!

File: torch/fx/experimental/symbolic_shapes.py (hunk 2 above)
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py:1758]: OK, let's punt.

**Comment by malfet** [comment]
@avikchaudhuri I've reverted this PR in https://github.com/pytorch/pytorch/commit/496212f4089c6898d4439f7205b20e108f95154c as it broke ExecuTorch internally, which you were are at the time you merged this PR, haven't you?

**Comment by avikchaudhuri** [comment]
@malfet Retrying because I couldn't repro the failures internally on latest trunk. The failures were supposed to be fixed by another diff in that train.

**Pull request closed without merging**